mdp/linear_flows_online.py

```diff
@@ -0,0 +1,658 @@
+        """
+
+        # if a single array is given, nodes are trained
+        # incrementally if it is a 2D array or block incrementally if it
+        # is a 3d array (num_blocks, block_size, dim).
+        if isinstance(data_iterables, numx.ndarray):
+            if data_iterables.ndim == 2:
+                data_iterables = data_iterables[:, mdp.numx.newaxis, :]
+            return data_iterables
+
+        # check it it is an iterable
+        if (data_iterables is not None) and (not hasattr(data_iterables, '__iter__')):
+            err = "data_iterable is not an iterable."
+            raise FlowException(err)
+
+        return data_iterables
+
+    def train(self, data_iterables):
+        """Train all trainable nodes in the flow.
+
+        'data_iterables' is a single iterable (including generator-type iterators if
+        the last node has no multiple training phases) that must return data
+        arrays to train nodes (so the data arrays are the 'x' for the nodes).
+        Note that the data arrays are processed by the nodes
+        which are in front of the node that gets trained,
+        so the data dimension must match the input dimension of the first node.
+
+        'data_iterables' can also be a 2D or a 3D numpy array. A 2D array trains
+        all the nodes incrementally, while a 3D array supports online training
+        in batches (=shape[1]).
+
+        'data_iterables' can also return a list or a tuple, where the first entry is
+        'x' and the rest are the required args for training all the nodes in
+        the flow (e.g. for supervised training).
+
+        (x, (node-0 args), (node-1 args), ..., (node-n args)) - args for n nodes
+
+        if say node-i does not require any args, the provided (node-i args) are ignored.
+        So, one can simply use None for the nodes that do not require args.
+
+        (x, (node-0 args), ..., None, ..., (node-n args)) - No args for the ith node.
+
+        """
+
+        data_iterables = self._train_check_iterables(data_iterables)
+
+        if self.verbose:
+            strn = [str(self.flow[i]) for i in xrange(len(self.flow))]
+            print("Training nodes %s simultaneously" % strn)
+        self._train_nodes(data_iterables)
+
+        # close training the terminal node only if it is a trainable Node
+        if not isinstance(self.flow[-1], mdp.OnlineNode):
+            self._close_last_node()
+
+    # private container methods
+
+    def _check_value_type_is_compatible(self, value):
+        # onlinenodes, trained and non-trainable nodes are compatible
+        if not isinstance(value, mdp.Node):
+            raise TypeError("flow item must be a Node instance and not %s" % type(value))
+        elif isinstance(value, mdp.OnlineNode):
+            pass
+        else:
+            # classic mdp Node
+            if value.is_training():
+                raise TypeError("flow item must either be an OnlineNode instance, a trained or a non-trainable Node.")
```

Comment:
I think there seems to be a confusion with the compatible node sequences for an OnlineFlow. I improved the docstrings and also renamed a method to make this clear. To answer your question let me first discuss the compatible node sequences:

OnlineFlow node sequence can contain
 (a) only OnlineNodes
        (Eg. [MovingAvgNode(), IncSFANode()],
 or
(b) a mix of OnlineNodes and trained/non-trainable Nodes
        (eg. [a fully trained PCANode, IncSFANode()] or [QuadraticExpansionNode(), IncSFANode()],
or
 (c) a mix of OnlineNodes/trained/non-trainable Nodes and a terminal trainable Node (but not an OnlineNode) whose training hasn't finished
        (eg. [IncSFANode(), QuadraticExpansionNode(), a partially or untrained SFANode]).

(a) is fairly obvious. (b) allows inclusion of non-trainable or pre-trained nodes. For example, a pretrained three layered hierarchical SFA node and a top incremental SFA node. (c) allows an uncommon combination of online preprocessing for training a batch node. 

The confusion could also stem from the method names I chose before:
"_check_value_type_is_compatible" (an interim compatibility check)
and 
"_check_compatibilitiy" (this is the actual one)

 I changed these to 
"_check_value_type_is_online_or_nontrainable_node" (renamed)
and 
"_check_compatibilitiy" (same as before).

I hope this clarifies your doubt. Thanks.